lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+const Ajv = require('ajv');
+const _ = require('lodash');
+const schema = require('../configSchema');
+const errorMessageBuilder = require('./ConfigSchemaHandler/errorMessageBuilder');
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = _.cloneDeep(schema);
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.package);
+    Object.freeze(this.schema.properties.layers);
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors);
+      for (const message of messages) {
+        switch (this.serverless.service.configValidationMode) {
+          case 'error':
+            throw new this.serverless.classes.Error(message);
```

Comment:
I think in a `for..of` loop we should gather all error messages, and after processing a loop, we should either log or throw a combined message.

The way it is now, on _error_ mode, we report just first error, which is not great UX wise.

Reply:
I agree this loop is unnecessary because the `messages` array contains only one element. I'd suggest once we gather feedback and we then refactor it for better UX.